### DISCUSSION TITLE
How best to implement a custom Option type

### DISCUSSION THREAD

**Discussion by markcatley** [jonasbb/serde_with#595] [Q&A] [answered]
Hi,

I have a CSV-like data format that has optional fields that sometimes are blank and sometimes have a fixed string in them ("--"). I've been trying to implement a custom type that implements DeserializeAs but I'm not really sure the best way to do it. I'd kinda like to "peek" into the field and check if it's empty or with that fixed string and then pass the deserializer to the type that's contained but I can't really figure out a good way to do that.

Thanks for your help.

**Comment by jonasbb** [comment, marked as answer]
Peeking is impossible with serde, because any `deserialize_*` method on `Deserializer` consumes the `Deserializer`. What serde does to simulate peeking is deserialize into a `Value`-like type, which can hold any data, and then deserialize from the `Value`, potentially multiple times. That strategy has some downsides though.

In your case you can deserialize a `String` or `Option<String>` and inspect that. You can turn the `String` into a `Deserializer` for example with [`StringDeserializer`](https://docs.rs/serde/latest/serde/de/value/struct.StringDeserializer.html).
The `StringDeserializer` is relativly strict and only provides `Strings`, whereas the original `Deserializer` might be more lenient and offer integers or booleans (I don't recall what csv can do exactly). If you need that you must either change your types to deserialize from `String`s, e.g., with [`DisplayFromStr`](https://docs.rs/serde_with/latest/serde_with/struct.DisplayFromStr.html) or make your buffering smarter.
You can check if the buffered `String` only contains numbers and then use a matching `*Deserializer` from `serde::de::value`.
